presto-main/src/main/java/com/facebook/presto/sql/analyzer/Windows.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.sql.tree.AstVisitor;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.Node;
+import com.facebook.presto.sql.tree.SortItem;
+import com.facebook.presto.sql.tree.Window;
+import com.facebook.presto.sql.tree.WindowDefinition;
+import com.facebook.presto.sql.tree.WindowFrame;
+import com.facebook.presto.sql.tree.WindowInline;
+import com.facebook.presto.sql.tree.WindowName;
+import com.facebook.presto.sql.tree.WindowSpecification;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+
+import static com.facebook.presto.sql.analyzer.SemanticErrorCode.INVALID_WINDOW_SPECIFICATION;
+import static com.google.common.base.Preconditions.checkArgument;
+
+final class Windows
+{
+    private Windows()
+    {
+    }
+
+    static Map<String, WindowSpecification> createWindowSpecificationMap(Node node, List<WindowDefinition> windowDefinitions, Function<String, WindowSpecification> windowSpecificationLookup)
+    {
+        Map<String, WindowSpecification> windowSpecifications = new HashMap<>();
```

Comment:
It's operating on the map as it's iterating through the definitions both to catch duplicates and to resolve references. `ImmutableMap.Builder` lacks `containsKey` and `get` which are used.